metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py

```diff
@@ -0,0 +1,341 @@
+            dashboard.workspace_id, "dashboards", dashboard.id
+        )
+
+    def get_report_users(self, workspace_id: str, report_id: str) -> List[User]:
+        return self._get_entity_users(workspace_id, "reports", report_id)
+
+    def get_reports(self, workspace: Workspace) -> List[Report]:
+        """
+        Fetch the report from PowerBi for the given Workspace
+        """
+        if workspace is None:
+            logger.info("workspace is None")
+            return []
+
+        reports: List[Report] = self._get_resolver().get_reports(workspace)
+
+        def fill_ownership() -> None:
+            if self.__config.extract_ownership is False:
+                logger.info(
+                    "Skipping user retrieval for report as extract_ownership is set to false"
+                )
+                return
+
+            for report in reports:
+                report.users = self.get_report_users(
+                    workspace_id=workspace.id, report_id=report.id
+                )
+
+        def fill_tags() -> None:
+            if self.__config.extract_endorsements_to_tags is False:
+                logger.info(
+                    "Skipping endorsements tags retrieval for report as extract_endorsements_to_tags is set to false"
+                )
+                return
+
+            for report in reports:
+                report.tags = workspace.report_endorsements.get(report.id, [])
+
+        fill_ownership()
+        fill_tags()
+
+        return reports
+
+    def get_workspaces(self) -> List[Workspace]:
+        groups: List[dict] = self._get_resolver().get_groups()
+        workspaces = [
+            Workspace(
+                id=workspace["id"],
+                name=workspace["name"],
+                datasets={},
+                dashboards=[],
+                reports=[],
+                report_endorsements={},
+                dashboard_endorsements={},
+                scan_result={},
+            )
+            for workspace in groups
+        ]
+        return workspaces
+
+    def _get_scan_result(self, workspace: Workspace) -> Any:
+        scan_id: Optional[str] = None
+        try:
+            scan_id = self.__admin_api_resolver.create_scan_job(
+                workspace_id=workspace.id
+            )
+        except requests.exceptions.HTTPError as e:
+            if data_resolver.is_permission_error(e):
+                logger.warning(
+                    "Dataset lineage can not be ingestion because this user does not have access to the PowerBI Admin "
+                    "API. "
+                )
+                return None
+            # raise error if other than 401 or 403
+            raise e
+
+        logger.info("Waiting for scan to complete")
+        if (
+            self.__admin_api_resolver.wait_for_scan_to_complete(
+                scan_id=scan_id, timeout=self.__config.scan_timeout
+            )
+            is False
+        ):
+            raise ValueError(
```

Comment:
What is "workspace detail"? 

What does it mean "Please increase scan_timeout to wait"????

Say "Please increase the scan_timeout configuration value to wait longer for the scan job to complete".

Reply:
done

Reply:
Let's not throw here. Please warn or report failure